### DISCUSSION TITLE
Request: Support vips_percent for float32 images

### DISCUSSION THREAD

**Discussion by CoenraadS** [libvips/libvips#3233] [General]
Hi

I'm trying to support the following usecase

1. Load 32bit float grayscale image
2. Clamp values to 1-99 percentile
3. Linear map via `scale`

However currently I cannot do step 2 because the documentation says `vips_percent () The function works for uchar and ushort images only.`

I cannot change the order of steps because I would lose too much information.

In my manual non-vips testing, I achieved this by doing sort + IQR

I'm aware I can get the std from stats(), but I'm not sure zscore method can replace IQR

I also tried 

```
const ushortIm = im.cast("ushort");
const lowerBound = ushortIm.percent(1);
const upperBound = ushortIm.percent(99);

im = im.less(lowerBound).ifthenelse(lowerBound, image);
im = im.more(upperBound).ifthenelse(upperBound, image);

im = im.scale();
```

But the lowerBound values were too different then operation directly on float itself, and the operation was too destructive (unless I did something else wrong)

**Comment by jcupitt** [comment]
Hi @CoenraadS,

For efficiency, the `percent` operator builds a histogram and searches that to find the percent value, so you need to map your image to 0-255 or 0-65535 (depending on the precision you want), run `precent`, then map the result back to your original image.

Something like:

```python
#!/usr/bin/python3

import sys
import pyvips

# test float image ... a 1000x1000 pixels, with a smooth left to right
# gradient in the range 0-1
image = pyvips.Image.grey(1000, 1000)

# make a uchar variant we can compute the percent on ... 0-1 in the image
# becomes 0-255 in this version
x = (image * 255).cast("uchar")
thresh = x.percent(44)

# thresh is on the scaled image, so we need to reverse the transform above to
# find the threshold on the original
thresh /= 255

# threshold the original image and save
(image > thresh).write_to_file(sys.argv[1])
```

Here the scaling is just `*255` then `/255`, but the exact scaling you'd like can vary a lot and needs some knowledge of the image properties (you might even like a non-linear rescaling), so libvips leaves this up to the user and doesn't try to automate it.

**Reply by CoenraadS** [reply]
Yes unfortunately this does not give me enough precision :(

What is the best way to access the float buffer directly so I can manually perform IQR?

**Reply by jcupitt** [reply]
You can use a 16-bit LUT as well, would that be accurate enough? What range accuracy do you need?

**Reply by jcupitt** [reply]
I guess you could do it in two passes: do a quick 8-bit hist of the whole range to find an estimate of the median and sigma, then do a second pass in 16 bits of just the (median - a bit) to (median + 0.7sigma) range to try to get a better estimate. Would that be enough? Sorting every pixel might be extremely slow.

**Reply by CoenraadS** [reply]
I looked at my data a bit better and figured actually I can just clamp the lowerbound to 0, and then set the upper bound to 3*sigma and all is good :)

**Reply by jcupitt** [reply]
Cool! Yes, large amounts of float data are always a pain to deal with :( Some pixel somewhere is always +Inf or NaN argh. Ints forever!
